MWPW-164660: [3-in-1][Milo] Implement a link converter for CTAs using an iFrame

mirafedas opened the pull request:
To display the content of the 3-in-1 modal, we replace the iFrame link from the  [checkout-link.xlsx](https://adobe.sharepoint.com/:x:/r/sites/adobecom/_layouts/15/doc2.aspx?sourcedoc=%7BF9C23F75-A3DC-4454-8BE7-A7FD7A39A9D7%7D&file=checkout-link.xlsx&action=default&mobileredirect=true&DefaultItemOpen=1) file with a link to the appropriate commerce page that includes the segmentation step for the given offer. This replacement occurs only if the link contains the `modal=twp`, `modal=crm`, or `modal=d2p` parameters.

In the `threeInOne.js` module, we handle the communication protocol to respond to events emitted by the commerce page rendered inside the iFrame, and render our custom loader.

A delay in loading the commerce page content is expected, as we cannot preload the page displayed within the iFrame (similar to the behavior of the upgrade modal).

Authoring:
For commerce CTAs (created in https://milo.adobe.com/tools/ost):
1. Checkout link must have the `workflowStep=segmentation` and `modal=true` parameters,
e.g. `https://milo.adobe.com/tools/ost?osi=L2C9cKHNNDaFtBVB6GVsyNI88RlyimSlzVfkMM2gH4A&type=checkoutUrl&text=free-trial&workflowStep=segmentation&modal=true`
2. In the [checkout-link.xlsx](https://adobe.sharepoint.com/:x:/r/sites/adobecom/_layouts/15/doc2.aspx?sourcedoc=%7BF9C23F75-A3DC-4454-8BE7-A7FD7A39A9D7%7D&file=checkout-link.xlsx&action=default&mobileredirect=true&DefaultItemOpen=1) file all URLs that should be replaced with the new commerce URL should have the `modal` parameter set. The value should be the modal type: `modal=twp`, `modal=crm`, or `modal=d2p`.
I added these values in the file and previewed it, but I have not published it yet.

Resolves: [MWPW-NUMBER](https://jira.corp.adobe.com/browse/MWPW-NUMBER)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/drafts/mirafedas/3in1-test-page?martech=off
- After: https://mwpw-164660-3-in-1-modal--milo--mirafedas.hlx.page/drafts/mirafedas/3in1-test-page?martech=off

yesil commented on libs/blocks/merch/threeInOne.js:
```diff
@@ -0,0 +1,98 @@
+    loading: 'lazy',
+    class: 'loading',
+  });
+  const pCircle = createTag('sp-progress-circle', { label: 'progress circle', indeterminate: true, size: 'l' });
```
I think if the loading doesn't finish in certain amount of time, we should kill the spinner and show some kind of error message.
to be checked with Nick.

yesil commented on test/blocks/merch/merch.test.js:
```diff
@@ -869,62 +855,6 @@ describe('Merch Block', () => {
     });
   });
 
-  describe('checkout link with optional params', async () => {
```
why remove?

yesil commented on libs/blocks/merch/threeInOne.js:
```diff
@@ -0,0 +1,98 @@
+  const content = createTag('div', { class: 'milo-iframe' });
+  const iframe = createTag('iframe', {
+    src: iframeUrl,
+    title: 'Three-in-one modal',
```
We should be using an a11y and SEO friendly localized title here IMO.

yesil commented on libs/blocks/merch/threeInOne.js:
```diff
@@ -0,0 +1,98 @@
+import { createTag, getConfig } from '../../utils/utils.js';
+
+const MSG_SUBTYPE = {
```
a unit test is welcome.

3ch023 commented:
@yesil @mirafedas modifying [checkout-link.xlsx](https://adobe.sharepoint.com/:x:/r/sites/adobecom/_layouts/15/doc2.aspx?sourcedoc=%7BF9C23F75-A3DC-4454-8BE7-A7FD7A39A9D7%7D&file=checkout-link.xlsx&action=default&mobileredirect=true&DefaultItemOpen=1) is a potential CSO bomb. at some moment GWP might do their change and publich the file thus reenabling 3-in-1 modals on prod before they are ready.
we probably should remove that change. or i would at least test on some very 'unnoticable' product and notify GWP and make sure they won't publish the file

3ch023 commented on libs/features/mas/src/external.js:
```diff
@@ -2,7 +2,6 @@
     CheckoutType,
     WorkflowStep,
-    buildCheckoutUrl,
 } from '@pandora/commerce-checkout-url-builder';
```
lets remove the whole `'@pandora/commerce-checkout-url-builder'` dependency and the .tgz package from internal folder?

3ch023 commented on libs/features/mas/src/buildCheckoutUrl.js:
```diff
@@ -0,0 +1,172 @@
+ * Allow-listing is required because some consumers of the library don't use Typescript so they can pass any object as
+ * 'checkoutData'. The library will ignore any unknown key in 'checkoutData'.
+ */
+const ALLOWED_KEYS = new Set([
```
i know we copied it from pandora, but later we probably can remove this validation.
but for now lets keep as is

3ch023 reviewed (changes requested):
small request to remove checkout builder package from dependencies.
rest looks good to me

narcis-radu commented on libs/blocks/merch/merch.js:
```diff
@@ -496,7 +497,13 @@ export async function openModal(e, url, offerType, hash, extraOptions) {
     modal = await openFragmentModal(fragmentPath, getModal);
   } else {
-    modal = await openExternalModal(url, getModal, extraOptions);
+    const isThreeInOneModal = [MODAL_TYPE_3_IN_1.CRM, MODAL_TYPE_3_IN_1.D2P, MODAL_TYPE_3_IN_1.TWP].includes(el?.getAttribute('data-modal-type')) && el?.href;
```
If we add additional modal types we would need to maintain the code in two places. Why not use `Object.values` to create an array from the values?

narcis-radu replied:
The constants are part of merch, but the code looks like it could be any type of modal, without a direct connection to merch. We should either a) make this more explicit, b) make the modal types more general so they can be used outside of `merch` scope.

narcis-radu commented on libs/blocks/merch/threeInOne.js:
```diff
@@ -0,0 +1,96 @@
+      document.querySelector('.three-in-one iframe')?.classList?.remove('loading');
+      document.querySelector('.three-in-one sp-theme')?.remove();
+      break;
+    case MSG_SUBTYPE.EXTERNAL:
```
I see the same action being executed for `MSG_SUBTYPE.EXTERNAL`, `MSG_SUBTYPE.SWITCH`, `MSG_SUBTYPE.RETURN_BACK`. Maybe the code can be refactored a bit here.

narcis-radu commented on libs/blocks/modal/modal.css:
```diff
@@ -359,6 +386,13 @@
     width: 80%;
   }
 
+  .dialog-modal.three-in-one {
```
I see `.dialog-modal.upgrade-flow-modal` has the exact properties, can we combine them somehow?

narcis-radu commented on libs/blocks/merch/merch.css:
```diff
@@ -27,6 +27,24 @@ a[is='checkout-link'].con-button > span {
+.three-in-one sp-theme {
+  width: max-content;
+  height: max-content;
+  left: 50%; 
```
Would this handle RTL accordingly?

narcis-radu commented on libs/blocks/merch/merch.js:
```diff
@@ -489,14 +490,28 @@ export async function openModal(e, url, offerType, hash, extraOptions) {
     window.location.hash = hash;
     window.addEventListener('milo:modal:closed', () => {
-      window.history.pushState({}, document.title, `#${prevHash}`);
+      if (prevHash === '') {
```
Why not do something like
```js
window.history.pushState({}, document.title, prevHash !== '' ? `#${prevHash}` : `${window.location.pathname}${window.location.search}`)
```

yesil reviewed (changes requested):
I wonder whether we should be using Milo modal to display the 3-in-1 or use SWC to render the modal from MAS directly.
Thus, the integration will be re-usable outside adobe.com.
@mirafedas  I don't know how much work is this, especially given the timelines.
this is just to start the reflection on it.
cc: @npeltier @3ch023

mirafedas commented on libs/blocks/merch/three-in-one.js:
```diff
@@ -0,0 +1,97 @@
+  const pCircle = createTag('sp-progress-circle', { label: 'progress circle', indeterminate: true, size: 'l' });
+  const theme = createTag('sp-theme', { theme: 'spectrum', color: 'light', scale: 'medium', dir: 'ltr' });
+  theme.append(pCircle);
+  content.append(theme);
```
I rewrote it to use .innerHTML, and regarding the dark theme - if support is needed, I can work on it in the scope of the follow-up tickets, because this has to be aligned with the commerce team, in the current task it is not mentioned, and it would be good to avoid adding more changes to this PR to get it merged before the RCP

mirafedas commented:
@yesil thank you for reviewing, I addressed all comments. Regarding the comment about rendering the modal as SWC directly from MAS - I would wait for the commerce team to develop their web component (as they mentioned at the meeting)

Axelcureno commented on libs/features/mas/src/checkout-mixin.js:
```diff
@@ -143,6 +148,44 @@ export function CheckoutMixin(Base) {
+          } catch (error) {
+            this.masElement.log?.error('Failed to add 3-in-1 modal parameters', error);
+          }
+        }
```
Should we add a `catch()` here?

mirafedas replied:
There is already a catch statement here :)

Axelcureno reviewed (approved):
LGTM, just one question and one recommendation, both trivial.

Roycethan commented:
@mirafedas  Plz review:
1) Business tab when clicked in below link
https://mwpw-164660-3-in-1-modal--milo--mirafedas.hlx.page/drafts/mirafedas/3in1-test-page?martech=off#mini-plans-web-cta-photoshop-card
directly navigated to https://commerce.adobe.com/store/email?items%5B0%5D%5Bid%5D=DAC628518DDD843D4D19A8BF7D8FF1FE&co=US&lang=en&cli=adobe_com in a new tab

Expectation as per in 3 in 1 page:   it stays in the same page :
https://commerce.adobe.com/store/segmentation?ms=COM&ot=TRIAL&pa=phsp_direct_individual&cli=adobe_com&ctx=if&co=US&lang=en&af=uc_new_user_iframe%2Cuc_new_system_close&ss=segmentation&items%5B0%5D%5Bid%5D=7C2BA37CAA6F683E829D5412A0BA982E&fps=f&ooid=E27CB5D79014ACAB6953B091CEA72228
( we can find this while clicking Business tab in https://commerce.adobe.com/store/segmentation?ms=COM&ot=TRIAL&pa=phsp_direct_individual&cli=adobe_com&ctx=if&co=US&lang=en&af=uc_new_user_iframe%2Cuc_new_system_close&ss=segmentation&items%5B0%5D%5Bid%5D=7C2BA37CAA6F683E829D5412A0BA982E&fps=f&ooid=E27CB5D79014ACAB6953B091CEA72228 )

mirafedas commented:
@Roycethan hi, could you please verify again?

Roycethan commented:
> MWPW-164660

Post CLI fix reviewed https://mwpw-164660-3-in-1-modal--milo--mirafedas.hlx.page/drafts/mirafedas/3in1-test-page?martech=off
and user now see below urls for each type in i-frame:

twp: https://commerce.adobe.com/store/segmentation?ms=COM&ot=TRIAL&pa=phsp_direct_individual&cli=mini_plans&ctx=if&co=US&lang=en&af=uc_new_user_iframe%2Cuc_new_system_close
D2P: https://commerce.adobe.com/store/segmentation?ms=COM&ot=BASE&pa=aeft_direct_individual&cli=mini_plans&ctx=if&co=US&lang=en&af=uc_new_user_iframe%2Cuc_new_system_close
CRM: https://commerce.adobe.com/store/segmentation?ms=COM&ot=BASE&pa=ccsn_direct_individual&cli=creative&ctx=if&co=US&lang=en&af=uc_segmentation_hide_tabs%2Cuc_new_user_iframe%2Cuc_new_system_close

Note: Loading time for content still takes a bit longer time , i think its already a known issue we know, spinner exist during load
